src/util/result.h

```diff
@@ -8,58 +8,256 @@
+protected:
+    std::unique_ptr<ErrorInfo<F>> m_info;
+
+    ErrorInfo<F>& Info() LIFETIMEBOUND
+    {
+        if (!m_info) m_info = std::make_unique<ErrorInfo<F>>();
+        return *m_info;
+    }
+
+    //! Value setter methods which do nothing because this ResultBase class is
+    //! specialized for type T=void, so there is no result value for it to hold.
+    //! The other ResultBase specialization below for non-void T overrides these
+    //! methods to manage the result value it contains.
+    void ConstructValue() {}
+    template <typename O>
+    void MoveValue(O&& other) {}
+    void DestroyValue() {}
+
+    //! Helper function to construct a new Result success value or failure value
+    //! using the arguments provided.
+    template <bool Failure, typename Result, typename... Args>
+    static void ConstructResult(Result& result, Args&&... args)
+    {
+        if constexpr (Failure) {
+            result.Info().failure.emplace(std::forward<Args>(args)...);
+        } else {
+            result.ConstructValue(std::forward<Args>(args)...);
+        }
+    }
 
-    std::variant<bilingual_str, T> m_variant;
+    //! ConstructResult() overload peeling off a util::Error constructor argument.
+    template <bool Failure, typename Result, typename... Args>
+    static void ConstructResult(Result& result, util::Error error, Args&&... args)
+    {
+        result.AddError(std::move(error.message));
+        ConstructResult</*Failure=*/true>(result, std::forward<Args>(args)...);
+    }
 
-    template <typename FT>
-    friend bilingual_str ErrorString(const Result<FT>& result);
+    //! ConstructResult() overload peeling off a util::Warning constructor argument.
+    template <bool Failure, typename Result, typename... Args>
+    static void ConstructResult(Result& result, util::Warning warning, Args&&... args)
+    {
+        result.AddWarning(std::move(warning.message));
+        ConstructResult<Failure>(result, std::forward<Args>(args)...);
+    }
+
+    //! ConstructResult() overload peeling off a util::MoveMessages constructor argument.
+    template <bool Failure, typename Result, typename R, typename... Args>
+    static void ConstructResult(Result& result, util::MoveMessages<R> messages, Args&&... args)
+    {
+        result.MoveMessages(std::move(messages.m_result));
+        ConstructResult<Failure>(result, std::forward<Args>(args)...);
+    }
+
+    //! Helper function to move success or failure values and any error and
+    //! warning messages from one result object to another. The two result
+    //! objects must have compatible success and failure types.
+    template <bool Constructed, typename DstResult, typename SrcResult>
+    static void MoveResult(DstResult& dst, SrcResult&& src)
+    {
+        if (Constructed) {
+            if (dst) {
+                dst.DestroyValue();
+            } else {
+                dst.m_info->failure.reset();
```

Comment:
This line is missing test coverage. It could be achieved with the following diff (though there might be a more elegant way to achieve this):
```diff
-util::Result<int, FnError> AccumulateFn(bool success)
+util::Result<int, FnError> AccumulateFn(bool success, bool init_success)
 {
-    util::Result<int, FnError> result;
+    util::Result<int, FnError> result = IntFailFn(0, init_success);
     util::Result<int> x = MultiWarnFn(1) >> result;
     BOOST_REQUIRE(x);
     util::Result<int> y = MultiWarnFn(2) >> result;
@@ -195,8 +195,10 @@ BOOST_AUTO_TEST_CASE(check_returned)
     ExpectFail(EnumFailFn(ERR2), Untranslated("enum fail."), ERR2);
     ExpectFail(ChainedFailFn(ERR1, 5), Untranslated("chained fail. enum fail. warn."), 5);
     ExpectSuccess(MultiWarnFn(3), Untranslated("warn 0. warn 1. warn 2."), 3);
-    ExpectSuccess(AccumulateFn(true), Untranslated("warn 0. warn 0. warn 1. int 3 warn."), 3);
-    ExpectFail(AccumulateFn(false), Untranslated("int 3 error. warn 0. warn 0. warn 1."), ERR1);
+    ExpectSuccess(AccumulateFn(true, true), Untranslated("int 0 warn. warn 0. warn 0. warn 1. int 3 warn."), 3);
+    ExpectFail(AccumulateFn(false, true), Untranslated("int 3 error. int 0 warn. warn 0. warn 0. warn 1."), ERR1);
+    ExpectSuccess(AccumulateFn(true, false), Untranslated("int 0 error. warn 0. warn 0. warn 1. int 3 warn."), 3);
+    ExpectFail(AccumulateFn(false, false), Untranslated("int 0 error. int 3 error. warn 0. warn 0. warn 1."), ERR1);
     ExpectSuccess(TruthyFalsyFn(0, true), {}, 0);
     ExpectFail(TruthyFalsyFn(0, false), Untranslated("failure value 0."), 0);
     ExpectSuccess(TruthyFalsyFn(1, true), {}, 1);
```